File: src/Modbot.My.Web/src/lib/serverUrl.ts
export const MAX_SERVER_URL_LENGTH = 2048

/**
 * The origin of a Modbot server's address, or null when it is not one.
 *
 * The same rule as my.modbot.co's own `Common/ServerUrl.cs`, so an address the browser saves is one
 * my.modbot.co would pass on in the same shape: absolute `https` only, no username or password, and
 * only the origin kept (`https://modbot.example/settings?x=1` becomes `https://modbot.example`).
 *
 * An `http` URL is refused rather than warned about: stored, it would downgrade every redirect
 * `/go` makes to it afterwards.
 */
export function normaliseServerUrl(raw: string | null | undefined): string | null {
  if (!raw) return null

  const text = raw.trim()
  if (!text || text.length > MAX_SERVER_URL_LENGTH) return null

  let url: URL
  try {
    url = new URL(text)
  } catch {
    return null
  }

  if (url.protocol !== 'https:') return null
  if (url.username || url.password) return null

  return url.origin
}

File: src/Modbot.Cloud/Common/ServerUrl.cs
using System.Diagnostics.CodeAnalysis;

namespace Modbot.Cloud.Common;

/// <summary>
/// The one rule for a Modbot server's address, shared by the registry and the selector's saves so
/// that both hold addresses in the same shape and can be compared.
/// </summary>
public static class ServerUrl
{
    public const int MaxLength = 2048;

    /// <summary>
    /// Accepts an absolute https URL with no credentials in it, and keeps only its origin:
    /// <c>https://modbot.example/settings?x=1</c> becomes <c>https://modbot.example</c>.
    /// </summary>
    public static bool TryNormalise(string? raw, [NotNullWhen(true)] out string? origin)
    {
        origin = null;

        if (string.IsNullOrWhiteSpace(raw) || raw.Length > MaxLength)
            return false;

        if (!Uri.TryCreate(raw.Trim(), UriKind.Absolute, out var uri) || uri.Scheme != Uri.UriSchemeHttps)
            return false;

        // A URL with a username or password in it is refused rather than stored.
        if (!string.IsNullOrEmpty(uri.UserInfo))
            return false;

        origin = uri.GetLeftPart(UriPartial.Authority);
        return true;
    }
}
